### PR TITLE
Add Userspace RCU recipe.

### PR DESCRIPTION
**Pull request by szmyd**
Specify library name and version:  **userspace-rcu/0.11.4**

Userspace RCU implements a library around the Linux Kernel RCU interface. It is used by benchmarking and metrics gathering libraries at eBay Inc. and a dependency for many of our Open Source projects.

---

- [X] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [X] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/userspace-rcu/all/test_package/test_package.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
+#include <benchmark/benchmark.h>
```
Comment (prince-chrismc) [inline recipes/userspace-rcu/all/test_package/test_package.cpp:1]: We usually try to keep the test_package simplified, is it possible to have a simple main that called a few key functions just to test the linkage?

File: recipes/userspace-rcu/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        elif self.options.model == "signal":
+            lib_name = 'liburcu-signal'
+        else:
+            lib_name = 'liburcu-bp'
+        if self.options.shared:
+            self.copy("*{}.dll".format(lib_name), dst="bin", keep_path=False)
+            self.copy("*{}.so*".format(lib_name), dst="lib", keep_path=False)
+            self.copy("*{}.dylib".format(lib_name), dst="lib", keep_path=False)
+        else:
+            self.copy("*{}.a".format(lib_name), dst="lib", keep_path=False)
```
Comment (prince-chrismc) [inline recipes/userspace-rcu/all/conanfile.py:83]: Is there as reason that only on lib at a time is linked again but all of them are copied?

**Review by prince-chrismc — COMMENTED** [review]
:wave: thanks for opening this PR and contributing this recipe

I had two small nits and two questsions, I am not familiar with this project so hopefully you can fill in the gaps

File: recipes/userspace-rcu/all/test_package/test_package.cpp (hunk 1 above)
Comment (szmyd) [reply recipes/userspace-rcu/all/test_package/test_package.cpp:1]: Sure! I copied this from our internal code as I'm not the original consumer for the package and "just grabbed something". Can definitely remove this dependency if that's ideal.

File: recipes/userspace-rcu/all/conanfile.py (hunk 2 above)
Comment (szmyd) [reply recipes/userspace-rcu/all/conanfile.py:83]: I'm not sure I follow; this is the package() function so while all library variants are built, only the one selected with the `model` option is copied into the package (and thus added to the cpp_info linker flags). Am I missing something?

File: recipes/userspace-rcu/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import os
+
+required_conan_version = ">=1.33.0"
+
+
+class UserspaceRCUConan(ConanFile):
+    name = "userspace-rcu"
+    homepage ="https://liburcu.org/"
+    description = "Userspace RCU (read-copy-update) library"
+    topics = ("conan", "urcu")
```
[the remark this answers is not in the document]
Comment (szmyd) [reply recipes/userspace-rcu/all/conanfile.py]: :+1:

File: recipes/userspace-rcu/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    url = "https://github.com/conan-io/conan-center-index"
+    license = "LGPLv2.1"
```
[the remark this answers is not in the document]
Comment (szmyd) [reply recipes/userspace-rcu/all/conanfile.py]: :+1:

File: recipes/userspace-rcu/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        elif self.options.model == "signal":
+            lib_name = 'liburcu-signal'
+        else:
+            lib_name = 'liburcu-bp'
+        if self.options.shared:
+            self.copy("*{}.dll".format(lib_name), dst="bin", keep_path=False)
+            self.copy("*{}.so*".format(lib_name), dst="lib", keep_path=False, symlinks=True)
+            self.copy("*{}.dylib".format(lib_name), dst="lib", keep_path=False)
+        else:
+            self.copy("*{}.a".format(lib_name), dst="lib", keep_path=False)
```
Comment (jgsogo) [inline recipes/userspace-rcu/all/conanfile.py:83]: Probably the `if/else` here is not needed in the build-system builds only the requested configuration (shared/static). Anyway, it is harmless 👍

File: recipes/userspace-rcu/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+        self.copy("*.h", src="{}/_build/include/urcu/".format(self._source_subfolder), dst="include/urcu", keep_path=False)
+        lib_name = ''
+        if self.options.model == "generic":
+            lib_name = 'liburcu'
+        elif self.options.model == "mb":
+            lib_name = 'liburcu-mb'
+        elif self.options.model == "signal":
+            lib_name = 'liburcu-signal'
+        else:
+            lib_name = 'liburcu-bp'
```
Comment (jgsogo) [inline recipes/userspace-rcu/all/conanfile.py:77]: I would have preferred something like the following, that resembles the same logic you implemented in the `package_info` method:

```suggestion
        lib_name = 'liburcu'
        if self.options.model == "mb":
            lib_name += '-mb'
        elif self.options.model == "signal":
            lib_name += '-signal'
        elif self.options.model == "bp":
            lib_name += '-bp'
```

**Review by jgsogo — APPROVED** [review]
I added some suggestions, but they are just suggestions.

**Comment by prince-chrismc** [comment]
Please do not force push 🙏 GitHub forces us to restart the review which is not fun!

File: recipes/userspace-rcu/all/conanfile.py (hunk 2 above)
Comment (prince-chrismc) [reply recipes/userspace-rcu/all/conanfile.py:83]: We try to respect the original projects layout as closely as possible this way existing users of the project can continue what they are doing and new consumers can just follow the upstream docs without Conan getting in the way.

Looking of the project what we'd ideally have is the the projects install and model the `pkg_config` targets https://github.com/urcu/userspace-rcu/blob/7c40d7f5b2139ec9151daf8410b6b934234220d5/configure.ac#L310-L316

This was, as the original author intended, consumers can pick which libs they link against

***

My suggestion is 
- deleting the `model` option
- call the upstream install method (noted https://github.com/conan-io/conan-center-index/pull/11595#discussion_r924209798)
- add components (should be a one to one with the libs) [how to docs](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md#autotools)

Comment (szmyd) [reply recipes/userspace-rcu/all/conanfile.py:83]: Ah, I see. Thanks for the clarification!

**Comment by szmyd** [comment]
@prince-chrismc I've switched the `option` to a set of `components` following some of the Wayland and other recipes as examples. What I don't see in any of them or in the docs is a way to enforce only one component is selected. For example, in test_package it appears to consume *all* components and add every library to the linker. Is there a way to avoid this and force consumers to explicitly choose *a* component?

**Review by prince-chrismc — APPROVED** [review]
Looks really good!

File: recipes/userspace-rcu/all/test_package/test_urcu_fork.c [hunk 7]
```diff
@@ -0,0 +1,214 @@
+/*
```
Comment (SSE4) [inline recipes/userspace-rcu/all/test_package/test_urcu_fork.c:1]: would be nice to reduce it, calling one function (e.g. `rcu_register_thread`) is enough to test we can compile and link against the library

**Review by SSE4 — APPROVED** [review]

**Pull request merged**